### PR TITLE
1-bit Adam v2

### PR DESCRIPTION
**Pull request by conglongli**
Authors: @awan-10 @conglongli @samyam @jeffra 

What's new:

1) NCCL-based implementation which provides better performance and usability compared to the MPI-based implementation.
2) Add support to momentum masks for those parameters with constant zero gradients during training.
3) Bug fixes (e.g., https://github.com/microsoft/DeepSpeed/issues/813).

### CODE REVIEW

File: deepspeed/runtime/engine.py [hunk 1]
```diff
@@ -622,8 +622,12 @@ def _configure_basic_optimizer(self, model_parameters):
             from deepspeed.ops.lamb import FusedLamb
             optimizer = FusedLamb(model_parameters, **optimizer_parameters)
         elif self.optimizer_name() == ONEBIT_ADAM_OPTIMIZER:
-            from deepspeed.runtime.fp16.onebit_adam import OnebitAdam
+            from deepspeed.runtime.fp16.onebit.onebitadam import OnebitAdam
```
Comment (awan-10) [inline deepspeed/runtime/engine.py:625]: the filename can remain adam instead of onebitadam, right? The class name OnebitAdam is good but let's keep the filename as adam.py instead of onebitadam.py. It looks redundant when we are inside the onebit folder.

File: deepspeed/runtime/fp16/onebit/onebitadam.py [hunk 2]
```diff
@@ -6,18 +6,14 @@
 import importlib
```
Comment (awan-10) [inline deepspeed/runtime/fp16/onebit/onebitadam.py:6]: This file should not exist at all. We should delete it. Can you please use git rm on this file to stop tracking it?

Comment (awan-10) [reply deepspeed/runtime/fp16/onebit/onebitadam.py:6]: Sorry, misspoke. This file is now named adam.py as I had thought so I got confused here.

File: deepspeed/runtime/fp16/onebit/onebitadam.py [hunk 3]
```diff
@@ -372,3 +251,31 @@ def step(self, closure=None, grads=None):
                 self.deepspeed.enable_backward_allreduce = False
 
         return loss
+
+    def load_state_dict(self, state_dict):
```
Comment (awan-10) [inline deepspeed/runtime/fp16/onebit/onebitadam.py:255]: this is the part I am not sure so Jeff/Samyam can review it closely.

**Comment by awan-10** [comment]
Thank you @conglongli. I left some comments above. Everything else looks good to me.

File: deepspeed/runtime/engine.py (hunk 1 above)
Comment (awan-10) [reply deepspeed/runtime/engine.py:625]: just clarifying: I want it to look like from ```deepspeed.runtime.fp16.onebit.adam import OnebitAdam``` instead of ```from deepspeed.runtime.fp16.onebit.onebitadam import OnebitAdam```

Comment (conglongli) [reply deepspeed/runtime/engine.py:625]: Agree. Will change it.

File: docs/_tutorials/onebit-adam.md [hunk 4]
```diff
@@ -240,7 +271,8 @@ Below is the DeepSpeed configuration file for running BERT-large pre-training wi
       "weight_decay": 0.01,
       "bias_correction": false,
       "freeze_step": 23000,
-      "cuda_aware": true
+      "cuda_aware": false,
```
Comment (awan-10) [inline docs/_tutorials/onebit-adam.md:274]: can be removed I guess?

Comment (conglongli) [reply docs/_tutorials/onebit-adam.md:274]: Agree, will do.

File: README.md [hunk 5]
```diff
@@ -32,6 +32,7 @@ information [here](https://innovation.microsoft.com/en-us/exploring-ai-at-scale)
 
 # News
 * [2021/03/08] [ZeRO-3 Offload: Scale your models to trillion parameters without code changes while leveraging both CPUs & GPUs](https://www.deepspeed.ai/news/2021/03/07/zero3-offload.html)
+* [2021/03/04] [1-bit Adam v2: NCCL-based implementation and more](https://www.deepspeed.ai/tutorials/onebit-adam/)
```
Comment (jeffra) [inline README.md:35]: Can we update this date and move it above z3?

File: docs/index.md [hunk 6]
```diff
@@ -29,6 +29,7 @@ information [here](https://innovation.microsoft.com/en-us/exploring-ai-at-scale)
 
 # What's New?
 * [2021/03/08] [ZeRO-3 Offload: Scale your models to trillion parameters without code changes while leveraging both CPUs & GPUs](https://www.deepspeed.ai/news/2021/03/07/zero3-offload.html)
+* [2021/03/04] [1-bit Adam v2: NCCL-based implementation and more](https://www.deepspeed.ai/tutorials/onebit-adam/)
```
Comment (jeffra) [inline docs/index.md:32]: same here, let's update the date and move it top

File: tests/onebit/test_mpi_backend.py [hunk 7]
```diff
@@ -4,26 +4,26 @@
...
+from deepspeed.runtime.comm.mpi import MpiBackend
 
 comm = MPI.COMM_WORLD
 size = comm.Get_size()
 rank = comm.Get_rank()
 
 #TODO: Detect the hostname we are running on automatically
 torch.distributed.init_process_group(backend='nccl',
-                                     init_method='tcp://worker-1:2245',
+                                     init_method='tcp://worker-0:2245',
```
Comment (jeffra) [inline tests/onebit/test_mpi_backend.py:16]: side note, you should be able to replace this with `deepspeed.init_distributed` (https://deepspeed.readthedocs.io/en/latest/initialize.html#distributed-initialization) it will use mpi4py to properly initialize torch.distributed.

File: tests/unit/test_onebit.py [hunk 8]
```diff
@@ -0,0 +1,373 @@
...
+if TORCH_MAJOR < 1 or TORCH_MINOR < 8:
+    pytest.skip("NCCL-based 1-bit compression requires torch 1.8 or higher",
+                allow_module_level=True)
+
+try:
+    from apex import amp
+    _amp_available = True
+except ImportError:
+    _amp_available = False
+amp_available = pytest.mark.skip(_amp_available, reason="apex/amp is not installed")
```
Comment (jeffra) [inline tests/unit/test_onebit.py:23]: these tests don't need `amp` do they? i assume we can remove this

File: tests/unit/test_onebit.py [hunk 9]
```diff
@@ -0,0 +1,373 @@
...
+        "train_batch_size": 2,
+        "steps_per_print": 1,
+        "optimizer": {
+            "type": "OneBitAdam",
+            "params": {
+                "lr": 0.00015,
+                "weight_decay": 0.01,
+                "freeze_step": 2,
+                "cuda_aware": False,
+                "comm_backend_name": "nccl"
```
Comment (jeffra) [inline tests/unit/test_onebit.py:37]: Can we update our json docs to describe the OneBitAdam parameters?

I would add a subsection under here: https://www.deepspeed.ai/docs/config-json/#optimizer-parameters called OneBitAdam and add a table with all the params OneBitAdam supports and their defaults.

File: docs/_tutorials/onebit-adam.md [hunk 10]
```diff
@@ -1,7 +1,15 @@
...
-title: "1-bit Adam: Up to 5x less communication volume and up to 2x faster training"
+title: "1-bit Adam: Up to 5x less communication volume and up to 3.4x faster training"
 ---
 
+**Note:**
+This tutorial is updated on 03/04/2021 to reflect the 1-bit Adam v2. Changes include: 1) NCCL-based implementation which provides better performance and usability compared to the MPI-based implementation. 2) Add support to momentum masks for those parameters with constant zero gradients during training. 3) Bug fixes. See details below.
+{: .notice--info}
+
+**Watch out!**
+1) The NCCL-based implementation requires PyTorch >= 1.8 and NCCL >= 2.8.3. See details below. 2) Although 1-bit Adam is compatible with both FP16 and FP32, currently we only verified the convergence under mixed precision/FP16 training. 3) Currently 1-bit Adam is not compatible with pipeline parallelism. 4) Frequent checkpoint loading could hurt 1-bit Adam's convergence. See details below.
```
Comment (jeffra) [inline docs/_tutorials/onebit-adam.md:10]: can we update this text to say we only see NCCL >= 2.8.3 needed in 64 gpu+ scenarios.

File: docs/_tutorials/onebit-adam.md [hunk 11]
```diff
@@ -25,9 +33,19 @@ git submodule update --init --recursive
...
 ```
 
-### Pre-requisites for 1-bit Adam
+### 1.2 Pre-requisites for 1-bit Adam
+
+#### 1.2.1 (New in v2) NCCL-based implementation
+
+In 1-bit Adam v2, we introduce a new system implementation for compressed communication using the NCCL backend of PyTorch distributed. This significantly improves the usability due to NCCL’s integration with PyTorch distributed. The performance of our new NCCL-based implementation is also better than our earlier MPI-based implementation for Ethernet-based systems and on-par for InfiniBand-based systems. Thus we highly recommend users to choose this implementation.
+
+**Watch out!**
```
Comment (jeffra) [inline docs/_tutorials/onebit-adam.md:42]: Can we update this text to say we only see NCCL >= 2.8.3 needed in 64 gpu+ scenarios.

File: deepspeed/runtime/fp16/onebit/adam.py [hunk 12]
```diff
@@ -6,18 +6,14 @@
...
-from mpi4py import MPI
-from deepspeed.runtime.custom_collectives import gather_cuda, gather_host, allgather_cuda, allgather_host
+from deepspeed.utils.logging import logger
 
 
 class OnebitAdam(torch.optim.Optimizer):
     """Implements the 1-bit Adam algorithm. Currently GPU-only.
-    For usage example please see, TODO DeepSpeed Tutorial
+    For usage example please see, https://www.deepspeed.ai/tutorials/onebit-adam/
     It has been proposed in APMSqueeze (https://arxiv.org/abs/2008.11343)
```
Comment (jeffra) [inline deepspeed/runtime/fp16/onebit/adam.py:17]: should this paper link be updated to https://arxiv.org/abs/2102.02888 ?

File: README.md (hunk 5 above)
Comment (conglongli) [reply README.md]: Resolved, thanks!

File: docs/index.md (hunk 6 above)
Comment (conglongli) [reply docs/index.md]: Resolved, thanks!

**Review by jeffra — CHANGES_REQUESTED** [review]
Left a few (mostly minor) comments. Looks great, thanks for all the hard work guys! :)

Can we also update this doc to point to the new class name for OneBitAdam?

https://github.com/microsoft/DeepSpeed/blob/master/docs/code-docs/source/optimizers.rst

File: deepspeed/runtime/fp16/onebit/onebitadam.py (hunk 3 above)
Comment (conglongli) [reply deepspeed/runtime/fp16/onebit/onebitadam.py]: Got approval from Samyam

File: tests/onebit/test_mpi_backend.py (hunk 7 above)
Comment (conglongli) [reply tests/onebit/test_mpi_backend.py]: Resolved, thanks!

File: tests/unit/test_onebit.py (hunk 9 above)
Comment (conglongli) [reply tests/unit/test_onebit.py:31]: Resolved, thanks for pointing to the json docs

**Comment by conglongli** [comment]
> Left a few (mostly minor) comments. Looks great, thanks for all the hard work guys! :)
> 
> Can we also update this doc to point to the new class name for OneBitAdam?
> 
> https://github.com/microsoft/DeepSpeed/blob/master/docs/code-docs/source/optimizers.rst

Thanks @jeffra for the review. Yes I updated the doc to point to the new class name for OneBitAdam.

**Review by jeffra — APPROVED** [review]
